Feature Promo

jmrcmendoza opened the pull request:
## Description

### Feature
Added Promo
  - Model
  - Data-Access
  - Entity
  - Use-Cases
  - Controller
  - Routes

jmrcmendoza committed:
feat: added promo model

jmrcmendoza committed:
feat: added promo data-access

jmrcmendoza committed:
feat: added promo use cases

select all promos
insert promo

jmrcmendoza committed:
feat: added promo controller

select all promo
insert promo

jmrcmendoza committed:
feat: added promo entity

jmrcmendoza committed:
feat: added promo route

jmrcmendoza committed:
feat: updated index

added promo route

jmrcmendoza committed:
feat: updated promo use casee

add select one promo

jmrcmendoza committed:
feat: updated promo controller

add select one promo

jmrcmendoza committed:
feat: updated promo routes

add select one promo

jmrcmendoza committed:
feat: updated promo use case

add update promo

jmrcmendoza committed:
feat: updated promo controller

add update promo

jmrcmendoza committed:
feat: updated promo routes

add update promo

jmrcmendoza committed:
feat: updated promo use case

added delete promo

jmrcmendoza committed:
feat: updated promo controller

added delete promo

jmrcmendoza committed:
feat: updated promo routes

added delete promo

ralphcasipe1 commented on koa_crud/src/models/promo.ts:
```diff
@@ -0,0 +1,80 @@
+  enabled: boolean;
+  status: Status;
+  minimumBalance: number | null;
+  requireMemberFields: [MemberFields];
```
Base on our use case it is awkward to check the array of objects.
I think one of the best alternative is to add enum MemberFields with values `EMAIL`, `REAL_NAME` and `BANK_ACCOUNT`.

and the `requiredMemberFields` should only accept this values.

ralphcasipe1 replied:
Can you please rename this to `requiredMemberFields`

ralphcasipe1 commented on koa_crud/src/models/promo.ts:
```diff
@@ -0,0 +1,80 @@
+      minimumBalance: {
+        type: Number,
+      },
+      requireMemberFields: [
+        {
+          realName: String,
+          email: String,
+          bankAccount: Number,
+        },
+      ],
```
It is easy to check the fields of `email`, `realName` and `bankAccount` if the `requiredMemberFields`  only accepts array of strings

ralphcasipe1 reviewed (changes requested):
You should also add in `MemberModel` the `email`, `realName` and `bankAccount` fields 

Let the `PromoModel.requireMemberFields` be the checker if those fields are already supplied

jmrcmendoza committed:
refactor: rename require member field

jmrcmendoza committed:
feat: updated member model

added email and bank account fields

jmrcmendoza committed:
feat: updated promo model

rename required member fields
updated MemberFields to enum
update requiredMemberFields type

jmrcmendoza committed:
feat: updated promo validations

jmrcmendoza committed:
feat: updated delete promo validation

jmrcmendoza committed:
refactor: fix import order

jmrcmendoza committed:
feat: updated promo status validation

ralphcasipe1 commented on koa_crud/src/models/promo.ts:
```diff
@@ -0,0 +1,76 @@
+  enabled: boolean;
+  status: Status;
+  minimumBalance: number | null;
+  requiredMemberFields: [MemberFields];
```
should `MemberFields[]`

jmrcmendoza committed:
refactor: update requiredMemberFields type
